### ISSUE TITLE
Spring Session and Dev Tools Cause ClassCastException

### ISSUE THREAD

**Issue by rwinch** [spring-projects/spring-boot#3805]
If Spring Session Redis and Spring Boot Dev Tools are being used and a user defined class is persisted into HttpSession a ClassCastException will occur when trying to load it.

The problem is that Spring Session uses Spring Data Redis to load the class on each request. Spring Data Redis eventually uses `DefaultDeserializer` which uses an `ObjectInputStream` to load the class from the system `ClassLoader`.

This means that the value restored from session has a class defined by the system `ClassLoader`. The application itself loads the same class from Spring Boot's `RestartClassLoader`. Since the two classes are loaded from different `ClassLoader`s they cannot be cast from one to the other.

One possible solution is for Spring Redis to use a mechanism to customize the `ObjectInputStream` `ClassLoader` to use the `ThreadLocal.getThread().getContextClassLoader()`. For example, it could use `ConfigurableObjectInputStream`

I believe one alternative is that if `DefaultDeserializer` were loaded using `RestartClassLoader`, then the `ObjectInputStream` should use the `RestartClassLoader`.

Related (possibly replaces this issue): [SPR-13409](https://jira.spring.io/browse/SPR-13409)

**Comment by snicoll** [comment]
Similar problem with other caches whose content survive the application refresh.

**Comment by rwinch** [comment]
@snicoll Would [SPR-13409](https://jira.spring.io/browse/SPR-13409) help with cache too?

**Comment by snicoll** [comment]
probably not. Each cache implementation has its own way of deserializing cached data but I guess we could maybe tune the config when devtools runs. I am not sure if that's a huge problem though, persistent caching in dev mode seems like a lousy option. 

We had an issue with our SpringOne demo that I need to dig a bit more. I'll create a separate issue if necessary. Thanks!

**Comment by philwebb** [comment]
Unfortunately I've not found a way to work around this type of issue without making changes in the things that call `ObjectInputStream`. For now we're probably going to need to leave this as a known limitation.

**Comment by binaryFallout** [comment]
So if you want Redis, it's not possible to use the dev-tools anymore? I'm facing the same error, which took me hours to find out (spring beginner).

**Comment by wilkinsona** [comment]
> So if you want Redis, it's not possible to use the dev-tools anymore

As things stand you can't use Spring Session backed by Redis and DevTools together.

>  I'm facing the same error, which took me hours to find out (spring beginner).

I'm sorry that you spent hours finding this out. It's certainly not the experience that we want beginners to have. This is mentioned in the [known limitations section of the documentation](http://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using-boot-devtools-known-restart-limitations), but perhaps you didn't see the description or it didn't mean much too you?

I'm wondering if we would be better failing fast when we detect the Spring Session, Redis, and DevTools combination with a message explaining that it won't work.

**Comment by binaryFallout** [comment]
Thanks for your reply. Yes, you're right. The documentation was the answer, but to be honest, if you start with spring there is so much you take notice from and try to remember. I don't think it's a doc update worth.

**Comment by rwinch** [comment]
This can be resolved for Spring Session and Spring Data Redis once [DATAREDIS-427](https://jira.spring.io/browse/DATAREDIS-427) is resolved. Perhaps a bump to @christophstrobl could make that happen?

A workaround for this can be found in [spring-state-securing-restful-apis](https://github.com/rwinch/spring-state-securing-restful-apis/blob/6e056b77c28964174825d8e0296e097555ac08e5/messages-session/src/main/java/sample/config/WorkAroundsConfig.java#L66).

**Comment by wilkinsona** [comment]
While we're waiting on some changes in Spring Data Redis and Spring Session, we could probably auto-configure this workaround in DevTools.

**Comment by eagle8625** [comment]
I refer to this link to resolve my problem: http://stackoverflow.com/questions/2591779/cast-across-classloader, however I don't know why one class is read into ObjectOutputStream then written into ObjectInputStream can change its classloader.

**Comment by raysuliteanu** [comment]
FYI same problem with Apache Ignite; add an object to an Ignite cache in the restartedMain and get from cache in another thread (embedded Tomcat nio thread) with launcher classloader and get a CCE. Don't use devtools and problem goes away. This is on Boot 1.5.4 and Ignite 2.0.0
